lib/progress.js

```diff
@@ -118,7 +118,7 @@ export function estimate(requestAsync, isBrowser) {
 			responseStream = webStreams.toNodeReadable(response.body);
 			reader = responseStream._reader;
 		} else {
-			responseStream = response.body;
+			responseStream = response.clone().body;
```

Comment:
I understand that this PR hopes to address balena-io/balena-cli#2152 , and that it implements suggestions in the README.md of `node-fetch`:
https://github.com/node-fetch/node-fetch#custom-highwatermark

But I do not understand why those suggestions / these changes (await in parallel for an original stream and its clone, or alternatively increase the high water mark to 1MiB) should make any difference. Generally (without digging in the implementation of `node-fetch`), "cloning things" has negative performance implications (time to copy data and additional memory usage), while increasing buffer sizes (from 16KiB to 1MiB) may have a positive time implication but a negative memory implication. But most importantly, such changes should only affect performance (a bit more/less time or memory usage), not correctness.

In other words, the app (web UI, balena CLI) could run a bit faster or slower, but it should not hang.

Have we tested to confirm that the changes in this PR actually make a difference to balena-io/balena-cli#2152 ? I think @klutchell got to a point of being able to consistently reproduce the issue (downloading hanging), so it would make sense to ask him to test this PR before it is merged. In this regard, this PR should be a draft and not ready for merging. Even as a draft, balena CI publishes new versions in the npm registry and the `package.json` of downstream modules can use: new versions get published to the npm 

```json
"balena-request": "11.4.2-increase-hw-8a33bc2be826c087916788f34198153a27240ec5",
```

(Hash obtained from https://www.npmjs.com/package/balena-request )

The README.md of `node-fetch` does not really explain why they recommend the following cloning:

```js
const response = await fetch('https://example.com');
const r1 = await response.clone();

const results = await Promise.all([response.json(), r1.text()]);
```

Looking at the lines above: what does `await response.clone()` do exactly? What is being cloned, data or metadata? Then await in parallel for both the original response and the cloned response. Why? I don't understand it. :-)

And it gets worse when the README goes on to say that _"If for some reason you don't like the solution above,"_ [indeed I don't, because it does not make sense] _"since 3.x you are able to modify the highWaterMark option."_ Regardless of v2.x vs v3.x, how can it be that increasing a buffer size from 16KiB to 1MiB is an equivalent alternative to awaiting for a cloned stream in parallel with the original stream?

What is really going on here? :-) I wonder if even the author/maintainer of `node-fetch` (or some of the polyfills) actually understood what is going on. If the suggestions in the `node-fetch` README actually "solve" anything, my gut feeling is that it is actually _masking_ some underlying bug by "usually" avoiding corner cases or racing conditions that trigger an underlying bug.

But I am also pragmatic and would be OK with merging this PR even if we don't understand exactly what is going on. However, let's at least confirm that this PR actually solves or works around balena-io/balena-cli#2152 .

Reply:
Yeah, I didn't really look too much into it, because it sounded like a really similar situation, and [cloning](https://github.com/balena-io-modules/balena-image-manager/pull/52) solved the issue for @klutchell , but I guess the [original issue](https://github.com/node-fetch/node-fetch/issues/386) that the [docs](https://github.com/node-fetch/node-fetch#custom-highwatermark) were talking about was that they were awaiting the full resolution of a chunked response which had been cloned, and the back pressure from the cloned stream was causing the response to *look* like it was hanging, even though, in reality, the TCP socket buffer in the kernel was just full and not sending the ACK's for the server to keep sending packets.

I was really just hoping for a "quick fix". This problem doesn't make any sense though. I don't understand why backpressure would have any effect at all on the problem, because the upstream still shouldn't be calling "read" until they have data or the TCP stream has ended. I am going to try to reproduce the error on a Balena fin, and I will check back.